File: diatrack-wfe/src/app/api/models/user-preferences.ts
import {MovingAverageParams} from "./moving-average-params";
import {HistogramProfileType} from "../../pages/dashboard/dashboard-settings/histogram-profiles";

export interface UserPreferences {
    treatment?: TreatmentPreferences;
    dashboard?: DashboardPreferences;
}

export interface TreatmentPreferences {
    bglUnit: BglUnit;
    timeFormat: TimeFormat;
    targetBglRange: {
        min: number;
        max: number;
    };
    bglLowThreshold: number;
}

export interface DashboardPreferences {
    bglStatsHistogram: {
        profileType: HistogramProfileType;
        plotHeight?: number;
        plotColour: PlotColour;
        movingAverage?: MovingAverageParams;
        activityLog: boolean;
        dataLabels: boolean;
    }
}

export enum BglUnit {
    MgDl = "MgDl",
    MmolL = "MmolL"
}

export function getBglUnitDisplayValue(bglUnit: BglUnit) {
    switch (bglUnit) {
        case BglUnit.MmolL:
            return 'mmol/L';
        case BglUnit.MgDl:
            return 'mg/dL';
    }
}

export enum TimeFormat {
    TwelveHour = 12,
    TwentyFourHour = 24
}

export enum PlotColour {
    Uniform = 'Uniform',
    ScaledByBgl = 'ScaledByBgl'
}

File: diatrack-api/Models/UserPreferences.cs
using Elasticsearch.Net;

namespace Diatrack.Models
{
    public class UserPreferences
    {
        public TreatmentPreferences Treatment { get; set; }

        public DashboardPreferences Dashboard { get; set; }
    }

    public class TreatmentPreferences
    {
        [StringEnum]
        public BglUnit? BglUnit { get; set; }

        [StringEnum]
        public TimeFormat? TimeFormat { get; set; }

        public Range TargetBglRange { get; set; }

        public float BglLowThreshold { get; set; }
    }

    public enum BglUnit
    {
        MgDl,
        MmolL
    }

    public enum TimeFormat
    {
        TwelveHour,
        TwentyFourHour
    }

    public class Range
    {
        public float? Min { get; set; }

        public float? Max { get; set; }
    }

    public class DashboardPreferences
    {
        public BglStatsHistogramPreferences BglStatsHistogram { get; set; }
    }

    public class BglStatsHistogramPreferences
    {
        public string ProfileType { get; set; }

        public int? PlotHeight { get; set; }

        [StringEnum]
        public PlotColour? PlotColour { get; set; }

        public MovingAverageParams MovingAverage { get; set; }

        public bool? ActivityLog { get; set; }

        public bool? DataLabels { get; set; }
    }

    public enum PlotColour
    {
        Uniform,
        ScaledByBgl
    }
}
